lang/java/reef-bridge-java/src/main/java/org/apache/reef/javabridge/ActiveContextBridgeFactory.java

```diff
@@ -0,0 +1,99 @@
+import org.apache.reef.driver.context.ActiveContext;
+import org.apache.reef.driver.context.ClosedContext;
+import org.apache.reef.driver.context.ContextBase;
+import org.apache.reef.tang.ClassHierarchy;
+import org.apache.reef.tang.formats.AvroConfigurationSerializer;
+
+import javax.inject.Inject;
+import java.util.HashMap;
+import java.util.Map;
+import java.util.logging.Logger;
+
+/**
+ * Factory for ActiveContextBridge instances.
+ * <p/>
+ * This class is thread safe because all its methods are synchronized. It should only be used within the bridge code.
+ */
+@DriverSide
+@ThreadSafe
+@Private
+public final class ActiveContextBridgeFactory {
+  private static final Logger LOG = Logger.getLogger(ActiveContextBridge.class.getName());
+
+  private final Map<String, ActiveContextBridge> activeContextBridgeCache = new HashMap<>();
+  private final ClassHierarchy clrClassHierarchy = Utilities.loadClassHierarchy(NativeInterop.CLASS_HIERARCHY_FILENAME);
+  private final AvroConfigurationSerializer configurationSerializer;
+
+  /**
+   * This is always instantiated via Tang.
+   *
+   * @param configurationSerializer passed to the ActiveContextBridge instances for configuration serialization.
+   */
+  @Inject
+  private ActiveContextBridgeFactory(final AvroConfigurationSerializer configurationSerializer) {
+    this.configurationSerializer = configurationSerializer;
+  }
+
+  /**
+   * Instantiates a new ActiveContextBridge or returns the one created earlier.
+   *
+   * @param context the context for which to return an ActiveContextBridge instance.
+   * @return a new ActiveContextBridge or returns the one created earlier.
+   */
+  public synchronized ActiveContextBridge getActiveContextBridge(final ActiveContext context) {
+    final String contextId = getUniqueContextId(context);
+
+    ActiveContextBridge result = activeContextBridgeCache.get(contextId);
+    if (null == result) {
+      result = new ActiveContextBridge(context, this.clrClassHierarchy, this.configurationSerializer);
+      activeContextBridgeCache.put(contextId, result);
+    }
+    return result;
+  }
+
+  /**
+   * Removes the ActiveContextBridge associated with the given Context from the cache.
+   *
+   * @param context the context whose ActiveContextBridge instance shall be removed from the cache.
+   */
+  public synchronized void ForgetContext(final ClosedContext context) {
+    final String contextId = getUniqueContextId(context);
+    final ActiveContextBridge removed = this.activeContextBridgeCache.remove(contextId);
+    if (null != removed) {
+      LOG.warning("Called with context previously unknown: " + contextId);
+    }
+  }
+
+  /**
+   * Constructs a unique ID for the given context by combining the evaluator and context ids.
+   *
+   * @param contextBase
+   * @return
+   */
+  private static String getUniqueContextId(final ContextBase contextBase) {
+    return contextBase.getEvaluatorId() + "/" + contextBase.getId();
```

Comment:
i was thinking about exactly the same solution, but then on a second thought feel there may still be some problem by combining evaluator id and context id:
technically contextId is named by user, and a user has the perfect right to choose the id of two contexts to be the same (if, say, he doesn't care about particular order of context), so this solution will depend on user being a good citizen, which is not something we can count on.